Asgard/Communications/Classes/CbusEventManager.cs

```diff
@@ -0,0 +1,246 @@
+                // TODO: set large fields to null
+                this.IsDisposed = true;
+            }
+        }
+
+        // // TODO: override finalizer only if 'Dispose(bool disposing)' has code to free unmanaged resources
+        // ~CbusEventManager()
+        // {
+        //     // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+        //     Dispose(disposing: false);
+        // }
+
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        #endregion
+
+        #region Methods
+
+        public bool? GetState(ushort nodeNumber, ushort eventNumber)
+        {
+            var key =
+                this.cbusEventStates.Keys
+                    .Where(n => !n.IsShort)
+                    .Where(n => n.NodeNumber == nodeNumber)
+                    .Where(n => n.EventNumber == eventNumber)
+                    .FirstOrDefault();
+            if (key is null) return null;
+            return this.cbusEventStates[key];
+        }
+
+        public bool? GetState(ushort eventNumber)
+        {
+            var key =
+                this.cbusEventStates.Keys
+                    .Where(n => n.IsShort)
+                    .Where(n => n.EventNumber == eventNumber)
+                    .FirstOrDefault();
+            if (key is null) return null;
+            return this.cbusEventStates[key];
+        }
+
+        public void RegisterCbusEvent<T>(ushort nodeNumber, ushort eventNumber, CbusEventCallback callback)
+            where T : class, ICbusAccessoryEvent
+        {
+            var key = CbusEventKey.Create<T>(nodeNumber, eventNumber);
+            this.cbusEventCallbacks[key] = callback;
+        }
+
+        #endregion
+
+        #region Support routines
+
+        private CbusEventKey? FindKey(ICbusAccessoryEvent cbusAccessoryEvent)
+        {
+            var result =
+                this.cbusEventCallbacks.Keys
+                    .Where(n => cbusAccessoryEvent.IsShortEvent == n.IsShort)
+                    .Where(n => n.IsShort || n.NodeNumber == (cbusAccessoryEvent as IHasNodeNumber)?.NodeNumber)
+                    .Where(n => (cbusAccessoryEvent.IsLongEvent && n.EventNumber == (cbusAccessoryEvent as IHasEventNumber)?.EventNumber) ||
+                                (cbusAccessoryEvent.IsShortEvent && n.EventNumber == (cbusAccessoryEvent as IHasDeviceNumber)?.DeviceNumber))
+                    .FirstOrDefault();
+            return result;
+        }
+
+        private CbusEventCallback? GetRegisteredCallback(ICbusAccessoryEvent cbusAccessoryEvent)
+        {
+            var key = FindKey(cbusAccessoryEvent);
+            if (key is null) return null;
+            return this.cbusEventCallbacks[key];
+        }
+
+        private void ProcessQueue()
```

Comment:
I'm wondering if this is just complicating things here at the moment - the callbacks (and the associated exception handling) are alrady inside `Task.Run`, may as well just deal with the exception fully there too

Reply:
I was thinking that more information may get spat out than just the exception. But I want the event handling processed as quickly as possible; the exception handling can be deferred if necessary.